src/main/scala/com/databricks/labs/overwatch/validation/DeploymentValidation.scala

```diff
@@ -0,0 +1,581 @@
+      case "Common_ETLStoragePrefix" => "ETL Storage Prefix should be common across the workspaces."
+      case "Common_ETLDatabase" => "Workspaces should have a common ETL Database Name."
+      case "Common_ConsumerDatabaseName" => "Workspaces should have a common Consumer Database Name."
+      case "Valid_Cloud_providers" => "Cloud provider can be either AWS or Azure."
+      case "NOTNULL_APIURL" => "API URL should not be empty."
+      case "NOTNULL_SecretScope" => "Secrete scope should not be empty."
+      case "NOTNULL_SecretKey_DBPAT" => "PAT key should not be empty."
+      case "Valid_PrimordialDate" => "Primordial Date should in yyyy-MM-dd format(Ex:2022-01-30) and should be less than current date."
+      case "Valid_MaxDays" => "Max Days should be a number."
+      case "APIURL_Connectivity" => "API URL should give some response with provided scope and key."
+      case "Validate_AuditLogPrefix" => "Folder with Primordial date should be present inside AuditLogPrefix."
+      case "Validate_EventHub" => "Consuming data from EventHub."
+      case "Valid_Excluded_Scopes" => "Excluded scope can be audit:sparkEvents:jobs:clusters:clusterEvents:notebooks:pools:accounts."
+      case "Storage_Access" => "ETL_STORAGE_PREFIX should have read,write and create access"
+    }
+  }
+
+  /**
+   * Performs mandatory validation before each deployment.
+   */
+  private[overwatch] def performMandatoryValidation = {
+    logger.log(Level.INFO, "Performing mandatory validation")
+    validateFileExistance
+    makeDataFrame
+    val groupedRuleSet = RuleSet(inputDataFrame, by = "deploymentId")
+    groupedRuleSet.add(validateEtlStorage)
+    validateRuleAndUpdateStatus(groupedRuleSet, true)
+    if (validationStatus.toDS().filter("validated==false").count > 0) { //Checks for failed validations
+      throw new BadConfigException(getSimpleMsg("Common_ETLStoragePrefix"))
+    }
+    storagePrefixAccessValidation(inputDataFrame.head(), true)
+  }
+
+  /**
+   * Entry point of the validation.
+   *
+   * @return
+   */
+  private[overwatch] def performValidation: Dataset[DeploymentValidationReport] = {
+    //Primary validation //
+    validateFileExistance
+    makeDataFrame
+    println("Parallelism :" + parallelism)
+
+    //csv data validation
+    val processingStartTime = System.currentTimeMillis();
+    val groupedRuleSet = RuleSet(inputDataFrame, by = "deploymentId")
+    groupedRuleSet.add(validateEtlStorage)
+    groupedRuleSet.add(validateETLDatabaseName)
+    groupedRuleSet.add(validateConsumerDatabaseName)
+    validateRuleAndUpdateStatus(groupedRuleSet, true)
+
+    val nonGroupedRuleSet = RuleSet(inputDataFrame)
+    nonGroupedRuleSet.add(validateCloud)
+    nonGroupedRuleSet.add(validateAPIURL)
+    nonGroupedRuleSet.add(validateSecretScope)
+    nonGroupedRuleSet.add(validateSecreteKeyDBPAT)
+    nonGroupedRuleSet.add(validatePrimordialDate)
+    nonGroupedRuleSet.add(validateMaxDays)
+    nonGroupedRuleSet.add(validateOWScope())
```

Comment:
create an Array of rules to be validated and pass into RuleSet instead of using .add.add.add...

Reply:
much better but can still be improved. Define the ruleSeq first and then use functional pattern to add it. It's minutia but makes it cleaner. 

```scala
val nonGroupedRuleSet = RuleSet(configDF)
    .add(nonGroupedRuleSet)
```

Reply:
done